oak_idl_gen_services/src/lib.rs

```diff
@@ -0,0 +1,383 @@
+///   request to the appropriate method on the underlying service implementation.
+/// - a trait named `Name`, with a method for each method defined in the macro, and an additional
+///   default method named `serve` which returns an instance of `NameServer`; the developer of a
+///   service would usually define a concrete struct and manually implement this trait for it.
+///
+/// For an input flatbuffer file with name `foo.fbs`, the generated Rust file will be located at
+/// `${OUT_DIR}/foo_services.rs`.
+pub fn compile_services(filename: &str) {
+    // Run flatc to generate the reflected schema in flatbuffer binary format.
+    let output = std::process::Command::new("flatc")
+        .args([
+            "--schema",
+            "--binary",
+            "-o",
+            &std::env::var("OUT_DIR").unwrap(),
+            filename,
+        ])
+        .output()
+        .unwrap();
+    if !output.status.success() {
+        eprintln!("flatc exit code: {}", output.status);
+        eprintln!(
+            "flatc stdout: {}",
+            std::str::from_utf8(&output.stdout).unwrap()
+        );
+        eprintln!(
+            "flatc stderr: {}",
+            std::str::from_utf8(&output.stderr).unwrap()
+        );
+        exit(1);
+    }
+    // We then read back the generated schema and parse it with the reflection flatbuffer schema.
+    // See https://google.github.io/flatbuffers/intermediate_representation.html.
+    let file_prefix = std::path::Path::new(filename)
+        .file_prefix()
+        .unwrap()
+        .to_str()
+        .unwrap();
+    let schema_file = format!("{}/{}.bfbs", std::env::var("OUT_DIR").unwrap(), file_prefix);
+    let schema_bytes = std::fs::read(&schema_file).unwrap();
+    let generated_schema = generate_from_bytes(&schema_bytes).unwrap();
+    let out_file = format!(
+        "{}/{}_services.rs",
+        std::env::var("OUT_DIR").unwrap(),
+        file_prefix
+    );
+    std::fs::write(&out_file, generated_schema).unwrap();
+}
+
+/// Generate the Rust file from the input flatbuffer file.
+///
+/// In general, each flatbuffer file may have zero or more `rpc_service` instances defined.
+fn generate_from_bytes(schema_bytes: &[u8]) -> anyhow::Result<String> {
+    let schema = reflection_generated::reflection::root_as_schema(schema_bytes)
+        .context("could not parse schema")?;
+    Ok(schema
+        .services()
+        .context("could not find any service definition")?
+        .iter()
+        .map(|s| generate_service(&s))
+        .collect::<Result<Vec<_>, _>>()?
+        .into_iter()
+        .intersperse("\n".to_string())
+        .collect())
+}
+
+/// Generate the Rust objects from the input [`Service`] instance, corresponding to an `rpc_service`
+/// entry.
+fn generate_service(service: &Service) -> anyhow::Result<String> {
+    let mut lines = Vec::new();
+    lines.extend(vec![
+        format!("// Generated file, do not edit."),
```

Comment:
I think we could add for which service the file is generated, how it is generated, and a link to how it might be used?

That might help with interpreting.

Related: do generated files need a license?